java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -148,51 +139,33 @@ class ArcHostContextParticle : AbstractArcHostParticle() {
         }
     }
 
-    data class NamedHandleConnection(
-        val handleName: String,
-        val handleConnection: HandleConnection
+    private suspend fun tryInstantiateParticle(
+        particleName: String,
+        location: String
+    ): Particle = instantiatedParticles.getOrPut(
+        particleName,
+        { instantiateParticle(ParticleIdentifier.from(location)) }
     )
-    data class ParticleConnection(val particleName: String, val connection: NamedHandleConnection)
-
-    private fun createHandlesMap(
-        handleConnections: Map<String, List<ParticleConnection>>,
-        it: ArcHostParticle_Particles,
-        arcId: String
-    ): Map<String, HandleConnection> {
-        return handleConnections[it.particleName]?.associateBy(
-            { it.connection.handleName },
-            { it.connection.handleConnection }
-        ) ?: throw IllegalArgumentException(
-            "Can't find handleConnection for ${it.particleName} in $arcId"
-        )
-    }
 
-    private fun createHandleConnection(
-        entity: ArcHostParticle_HandleConnections,
+    private suspend fun createHandlesMap(
         arcId: String,
-        hostId: String,
-        instantiatedParticles: Map<String, Particle>
-    ): NamedHandleConnection {
-        return NamedHandleConnection(
-            entity.handleName,
-            HandleConnection(
-                StorageKeyParser.parse(entity.storageKey),
-                HandleMode.valueOf(entity.mode),
-                fromTag(
-                    arcId,
-                    requireNotNull(instantiatedParticles[entity.particleName]) {
-                        "${entity.particleName} not instantiable for $arcId and $hostId"
-                    },
-                    entity.type,
-                    entity.handleName
-                ), entity.ttl.let { num ->
-                    if (num != Ttl.TTL_INFINITE) Ttl.Minutes(
-                        num.toInt()
-                    ) else Ttl.Infinite
-                }
+        particleName: String,
+        particle: Particle,
+        handles: Set<Reference<ArcHostContextParticle_HandleConnections>>
+    ) = handles.map { ref ->
+            ref.dereference()
+        }.map {
+        requireNotNull(it) {
+            "HandleConnection couldn't be dereferenced for arcId $arcId, particle $particleName"
```

Comment:
Same comment here, you could pull the requireNotNull up to live with the `dereference` call